### PR TITLE
Add `torch.put_along_dim` and `torch.put_along_dim_` like `np.put_along_axis`

### LINKED ISSUE #120209: `torch.put_along_dim` and `torch.put_along_dim_` like `np.put_along_axis`
### 🚀 The feature, motivation and pitch

Much like `torch.take_along_dim` was added as a wrapper around `torch.gather` with broadcasting logic in #52833 to match `np.take_along_axis`, it would be nice to have wrappers around `torch.scatter_` and `torch.scatter` that "are designed to work with [...] [f]unctions that return indices along a dimension, like [`torch.argmax()`](https://pytorch.org/docs/stable/generated/torch.argmax.html#torch.argmax) and [`torch.argsort()`](https://pytorch.org/docs/stable/generated/torch.argsort.html#torch.argsort)," as the [documentation for `take_along_dim`](https://pytorch.org/docs/stable/generated/torch.take_along_dim.html) indicates.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by arunppsg**
Fixes #120209

cc @albanD

### CODE REVIEW

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp [hunk 1]
```diff
@@ -2147,6 +2147,53 @@ Tensor& take_along_dim_out(const Tensor& self, const Tensor& indices, std::optio
...
+    const Tensor& values,
+    c10::optional<int64_t> opt_dim) {
+  checkDevice("torch.put_along_dim():", {self, indices, values}, self.device());
+  if (opt_dim.has_value()) {
+    TORCH_CHECK(opt_dim.has_value(), "None dim is not supported.");
+    auto [new_indices, new_values] = _put_along_dim_helper(self, indices, values, opt_dim.value());
+    return at::scatter(self, opt_dim.value(), new_indices, new_values);
+  } else {
+    TORCH_CHECK(indices.dim() == 1, "when axis=None, `indices` must have a single dimension");
+    auto result = at::scatter(self.flatten(), 0, indices, values);
```
Comment (lezcano) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:2179]: Actually, can we use  https://pytorch.org/docs/stable/generated/torch.Tensor.put_.html here to avoid calling `flatten` (sparing us a copy)?

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp [hunk 2]
```diff
@@ -2147,6 +2147,53 @@ Tensor& take_along_dim_out(const Tensor& self, const Tensor& indices, std::optio
...
+}
+
+Tensor put_along_dim(
+    const Tensor& self,
+    const Tensor& indices,
+    const Tensor& values,
+    c10::optional<int64_t> opt_dim) {
+  checkDevice("torch.put_along_dim():", {self, indices, values}, self.device());
+  if (opt_dim.has_value()) {
+    TORCH_CHECK(opt_dim.has_value(), "None dim is not supported.");
```
Comment (lezcano) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:2174]: dead code?

File: test/test_indexing.py [hunk 3]
```diff
@@ -1558,6 +1558,31 @@ def test_take_along_dim_invalid(self, device, dtype):
...
+        def _test_against_numpy(t, indices, values, dim):
+            actual = torch.put_along_dim(t, indices, values, dim)
+            t_np = t.cpu().numpy()
+            indices_np = indices.cpu().numpy()
+            values_np = values.cpu().numpy()
+            # performs operation in-place
+            np.put_along_axis(t_np, indices_np, values_np, dim)
+            self.assertEqual(actual, t_np, atol=0, rtol=0)
+
+        t = torch.tensor([[10, 30, 20], [60, 40, 50]], dtype=dtype)
```
Comment (lezcano) [inline test/test_indexing.py:1572]: can you test the broadcasting behaviour between the different tensors, for both `dim` specified and `=None`?

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp [hunk 4]
```diff
@@ -2147,6 +2147,53 @@ Tensor& take_along_dim_out(const Tensor& self, const Tensor& indices, std::optio
...
+
+inline std::tuple<Tensor, Tensor> _put_along_dim_helper(const Tensor& self, const Tensor& indices, const Tensor& values, int64_t dim) {
+  auto new_indices_sizes = DimVector(indices.sizes());
+  for (int i = 0; i < indices.dim(); i++) {
+    if (i == dim) {
+      new_indices_sizes[i] = indices.size(i);
+    } else {
+      new_indices_sizes[i] = self.size(i);
+    }
+  }
```
Comment (lezcano) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:2160]:
```suggestion
  auto new_indices_sizes = DimVector(self.sizes());
  new_indices_sizes[dim] = indices.size(i);
```

**Review by lezcano — COMMENTED** [review]
This looks quite good!

Now, rather than adding a one-off test, could you add an OpInfo within `common_methods_invocations.py`. You can even specify a NumPy reference there and a test will test your implementation against it.

Make sure all the relevant tests in `test_ops*.py` pass locally before submitting :)

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp (hunk 4 above)
Comment (arunppsg) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: updated it in [b5a43f5](https://github.com/pytorch/pytorch/pull/125601/commits/b5a43f59ce6f1082bdc3868949be79c337f159bd)

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp (hunk 1 above)
Comment (arunppsg) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp]: fixes in [3f06685](https://github.com/pytorch/pytorch/pull/125601/commits/3f06685be5088d1405d507da3c242f0f11e5bd66)

**Comment by arunppsg** [comment]
Made some of the change and work in progress for OpInfo tests.

**Comment by lezcano** [comment]
changes so far look good. Thank you!

**Comment by arunppsg** [comment]
Added simple OpInfo test, though one test is failing: `test_non_standard_bool_values_put_along_dim_cpu_bool` - working on debugging it.
